### PR TITLE
Logan/router engine multi v2

### PR DESCRIPTION
**Pull request by logan-markewich**
# Description

This PR adds to main features
1. Multi-Select support to the router query engine
2. OpenAIPydanticProgram support for the selectors (rather than structured json parsing)

The multi-select works by gathering response from all sub-indexes, and then aggregating a final answer with a list index if there are more than 1 responses.

The OpennAIPydantic selectors are pretty basic, and mostly just needed specialized prompts to encourage the model to use the function call api (There will be a feature soon to force it to call the function, but langchain hasn't implemented it yet I think?)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# TODO:

- [x] create a good example notebook/usecase for the multi-router

### CODE REVIEW

File: llama_index/selectors/pydantic_selectors.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+        # parse output
+        return _pydantic_output_to_selector_result(prediction)
+
+    async def _aselect(
+        self, choices: Sequence[ToolMetadata], query: QueryBundle
+    ) -> SelectorResult:
+        raise NotImplementedError(
+            "Async selection not supported for Pydantic Selectors."
+        )
```
Comment (Disiok) [inline llama_index/selectors/pydantic_selectors.py:145]: Time to implement the `acall` in `BasePydanticProgram` class?

**Review by Disiok — APPROVED** [review]

File: llama_index/selectors/pydantic_selectors.py (hunk 1 above)
Comment (logan-markewich) [reply llama_index/selectors/pydantic_selectors.py:145]: Might be time! 😉

File: llama_index/selectors/pydantic_selectors.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    index: int
+    reason: str
+
+
+class MultiSelection(BaseModel):
+    """A multi-selection of choices."""
+
+    selections: List[SingleSelection]
+
+
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/selectors/pydantic_selectors.py:31]: Yea agreed. Made this process a little less redundant

File: llama_index/query_engine/router_query_engine.py [hunk 3]
```diff
@@ -49,40 +53,129 @@ def __init__(
...
+        response_docs = []
+        for response in responses:
+            if isinstance(response, StreamingResponse):
+                response_docs.append(Document(response.get_response().response))
+            else:
+                response_docs.append(Document(response.response))
+
+        summary_index = ListIndex.from_documents(response_docs)
+
+        query_engine = summary_index.as_query_engine(response_mode="tree_summarize")
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/query_engine/router_query_engine.py:84]: Added anyways!

**Pull request merged**